compiler-core/src/type_/name.rs

```diff
@@ -0,0 +1,81 @@
+use ecow::EcoString;
+
+use crate::ast::{ArgNames, SrcSpan};
+
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+pub trait NameChecker {
```

Comment:
We don't define new traits, please remove this.

Could you move all the logic for deciding whether names are valid elsewhere- it's not specific to types so it should go in the type_ module namespace.